File: test_connection.js
#!/usr/bin/env node
/**
 * Connection Test Script for Trading Bot
 * Tests connectivity to Charles Schwab and Robinhood brokers
 */

// Load environment variables from .env file
const fs = require('fs');
const path = require('path');

// Determine environment (defaults to 'local')
const env = (process.env.ENVIRONMENT || 'local').toLowerCase();
const envFile = `.env.${env}`;

// Load environment-specific .env file
if (fs.existsSync(envFile)) {
    require('dotenv').config({ path: envFile });
    console.log(`✓ Loaded environment from ${envFile}`);
} else {
    // Fallback to .env if environment-specific file doesn't exist
    require('dotenv').config();
    console.log('✓ Loaded environment from .env');
}

// Import broker classes from trading_bot
const { SchwabBroker, RobinhoodBroker } = require('./trading_bot');

// Broker configuration
const BROKERS = {
    schwab: {
        enabled: process.env.SCHWAB_ENABLED === 'true',
        apiKey: process.env.SCHWAB_API_KEY || '',
        appSecret: process.env.SCHWAB_APP_SECRET || '',
        callbackUrl: process.env.SCHWAB_CALLBACK_URL || 'https://127.0.0.1:8182/',
        tokenPath: process.env.SCHWAB_TOKEN_PATH || './schwab_token.json',
        accountId: process.env.SCHWAB_ACCOUNT_ID || '',
        baseUrl: process.env.SCHWAB_BASE_URL || ''
    },
    robinhood: {
        enabled: process.env.ROBINHOOD_ENABLED === 'true',
        username: process.env.ROBINHOOD_USERNAME || '',
        password: process.env.ROBINHOOD_PASSWORD || '',
        mfaCode: process.env.ROBINHOOD_MFA_CODE || '',
        apiKey: process.env.ROBINHOOD_API_KEY || '',
        deviceToken: process.env.ROBINHOOD_DEVICE_TOKEN || ''
    }
};

// Color codes for terminal output
const Colors = {
    GREEN: '\x1b[32m',
    RED: '\x1b[31m',
    YELLOW: '\x1b[33m',
    BLUE: '\x1b[34m',
    RESET: '\x1b[0m',
    BOLD: '\x1b[1m'
};

function printSuccess(message) {
    console.log(`${Colors.GREEN}✓ ${message}${Colors.RESET}`);
}

function printError(message) {
    console.log(`${Colors.RED}✗ ${message}${Colors.RESET}`);
}

function printWarning(message) {
    console.log(`${Colors.YELLOW}⚠ ${message}${Colors.RESET}`);
}

function printInfo(message) {
    console.log(`${Colors.BLUE}ℹ ${message}${Colors.RESET}`);
}

function printHeader(message) {
    console.log(`\n${Colors.BOLD}${Colors.BLUE}${'='.repeat(60)}${Colors.RESET}`);
    console.log(`${Colors.BOLD}${Colors.BLUE}${message}${Colors.RESET}`);
    console.log(`${Colors.BOLD}${Colors.BLUE}${'='.repeat(60)}${Colors.RESET}\n`);
}

async function testSchwabConnection() {
    printHeader('Testing Charles Schwab Connection');
    
    const config = BROKERS.schwab;
    
    // Check if enabled
    if (!config.enabled) {
        printWarning('Schwab is not enabled in configuration');
        printInfo('Set SCHWAB_ENABLED=true in your .env file to test');
        return { success: false, details: {} };
    }
    
    // Check credentials
    if (!config.apiKey || !config.appSecret) {
        printError('Schwab API credentials not configured');
        printInfo('Set SCHWAB_API_KEY and SCHWAB_APP_SECRET in your .env file');
        return { success: false, details: {} };
    }
    
    printInfo(`API Key: ${config.apiKey.substring(0, 10)}... (hidden)`);
    printInfo(`Callback URL: ${config.callbackUrl}`);
    printInfo(`Token Path: ${config.tokenPath}`);
    
    // Test connection
    try {
        const broker = new SchwabBroker(config);
        printInfo('Attempting to connect to Charles Schwab API...');
        
        const connected = await broker.connect();
        if (connected) {
            printSuccess('Successfully connected to Charles Schwab API');
            
            // Test account info retrieval
            printInfo('Testing account information retrieval...');
            const accountInfo = await broker.getAccountInfo();
            
            if (accountInfo && Object.keys(accountInfo).length > 0) {
                printSuccess('Successfully retrieved account information');
                const accountId = accountInfo.accountId || 'N/A';
                const accountType = accountInfo.type || 'N/A';
                printInfo(`  Account ID: ${accountId}`);
                printInfo(`  Account Type: ${accountType}`);
                
                // Test positions retrieval
                printInfo('Testing positions retrieval...');
                const positions = await broker.getPositions();
                printSuccess(`Successfully retrieved ${positions.length} positions`);
                
                // Test market data retrieval
                printInfo('Testing market data retrieval (SPY)...');
                const { price, indicators } = await broker.getMarketData('SPY');
                if (price > 0) {
                    printSuccess(`Successfully retrieved market data for SPY: $${price.toFixed(2)}`);
                    printInfo(`  Bid: $${(indicators.bid || 0).toFixed(2)}`);
                    printInfo(`  Ask: $${(indicators.ask || 0).toFixed(2)}`);
                    printInfo(`  Volume: ${(indicators.volume || 0).toLocaleString()}`);
                } else {
                    printWarning('Market data retrieval returned no price');
                }
                
                return {
                    success: true,
                    details: {
                        accountId,
                        accountType,
                        positionsCount: positions.length,
                        marketDataWorking: price > 0
                    }
                };
            } else {
                printWarning('Connected but could not retrieve account information');
                return { success: true, details: { accountInfo: false } };
            }
        } else {
            printError('Failed to connect to Charles Schwab API');
            return { success: false, details: {} };
        }
    } catch (error) {
        printError(`Error testing Schwab connection: ${error.message}`);
        printInfo('Full error details:');
        console.error(error);
        return { success: false, details: {} };
    }
}

async function testRobinhoodConnection() {
    printHeader('Testing Robinhood Connection');
    
    const config = BROKERS.robinhood;
    
    // Check if enabled
    if (!config.enabled) {
        printWarning('Robinhood is not enabled in configuration');
        printInfo('Set ROBINHOOD_ENABLED=true in your .env file to test');
        return { success: false, details: {} };
    }
    
    // Check credentials
    if (!config.username || !config.password) {
        printError('Robinhood credentials not configured');
        printInfo('Set ROBINHOOD_USERNAME and ROBINHOOD_PASSWORD in your .env file');
        return { success: false, details: {} };
    }
    
    printInfo(`Username: ${config.username}`);
    printInfo('Password: *** (hidden)');
    if (config.mfaCode) {
        printInfo('MFA Code: *** (hidden)');
    }
    
    // Test connection
    try {
        const broker = new RobinhoodBroker(config);
        printInfo('Attempting to connect to Robinhood API...');
        
        const connected = await broker.connect();
        if (connected) {
            printSuccess('Successfully connected to Robinhood API');
            
            // Test account info retrieval
            printInfo('Testing account information retrieval...');
            const accountInfo = await broker.getAccountInfo();
            
            if (accountInfo && Object.keys(accountInfo).length > 0) {
                printSuccess('Successfully retrieved account information');
                const username = accountInfo.username || 'N/A';
                printInfo(`  Username: ${username}`);
                
                // Test positions retrieval
                printInfo('Testing positions retrieval...');
                const positions = await broker.getPositions();
                printSuccess(`Successfully retrieved ${positions.length} positions`);
                
                // Test market data retrieval
                printInfo('Testing market data retrieval (SPY)...');
                const { price, indicators } = await broker.getMarketData('SPY');
                if (price > 0) {
                    printSuccess(`Successfully retrieved market data for SPY: $${price.toFixed(2)}`);
                    printInfo(`  Bid: $${(indicators.bid || 0).toFixed(2)}`);
                    printInfo(`  Ask: $${(indicators.ask || 0).toFixed(2)}`);
                    printInfo(`  Volume: ${(indicators.volume || 0).toLocaleString()}`);
                } else {
                    printWarning('Market data retrieval returned no price');
                }
                
                return {
                    success: true,
                    details: {
                        username,
                        positionsCount: positions.length,
                        marketDataWorking: price > 0
                    }
                };
            } else {
                printWarning('Connected but could not retrieve account information');
                return { success: true, details: { accountInfo: false } };
            }
        } else {
            printError('Failed to connect to Robinhood API');
            printInfo('Check your username, password, and MFA code if enabled');
            return { success: false, details: {} };
        }
    } catch (error) {
        printError(`Error testing Robinhood connection: ${error.message}`);
        printInfo('Full error details:');
        console.error(error);
        return { success: false, details: {} };
    }
}

async function main() {
    printHeader('Trading Bot Connection Test');
    
    const env = (process.env.ENVIRONMENT || 'local').toLowerCase();
    printInfo(`Environment: ${env}`);
    printInfo('Testing brokers based on configuration...\n');
    
    const results = {
        schwab: { success: false, details: {} },
        robinhood: { success: false, details: {} }
    };
    
    // Test Schwab
    const schwabResult = await testSchwabConnection();
    results.schwab = schwabResult;
    
    // Test Robinhood
    const robinhoodResult = await testRobinhoodConnection();
    results.robinhood = robinhoodResult;
    
    // Print summary
    printHeader('Test Summary');
    
    if (results.schwab.success) {
        printSuccess('Charles Schwab: CONNECTED ✓');
        if (results.schwab.details && Object.keys(results.schwab.details).length > 0) {
            const details = results.schwab.details;
            printInfo(`  Account ID: ${details.accountId || 'N/A'}`);
            printInfo(`  Positions: ${details.positionsCount || 0}`);
            printInfo(`  Market Data: ${details.marketDataWorking ? 'Working' : 'Not Working'}`);
        }
    } else {
        if (BROKERS.schwab.enabled) {
            printError('Charles Schwab: FAILED ✗');
        } else {
            printWarning('Charles Schwab: DISABLED (not configured)');
        }
    }
    
    if (results.robinhood.success) {
        printSuccess('Robinhood: CONNECTED ✓');
        if (results.robinhood.details && Object.keys(results.robinhood.details).length > 0) {
            const details = results.robinhood.details;
            printInfo(`  Username: ${details.username || 'N/A'}`);
            printInfo(`  Positions: ${details.positionsCount || 0}`);
            printInfo(`  Market Data: ${details.marketDataWorking ? 'Working' : 'Not Working'}`);
        }
    } else {
        if (BROKERS.robinhood.enabled) {
            printError('Robinhood: FAILED ✗');
        } else {
            printWarning('Robinhood: DISABLED (not configured)');
        }
    }
    
    // Exit code
    if (results.schwab.success || results.robinhood.success) {
        printSuccess('\n✓ At least one broker connection successful!');
        process.exit(0);
    } else {
        printError('\n✗ No broker connections successful. Please check your configuration.');
        process.exit(1);
    }
}

// Run tests
main().catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
});

File: test_connection.py
#!/usr/bin/env python3
"""
Connection Test Script for Trading Bot
Tests connectivity to Charles Schwab and Robinhood brokers
"""

import os
import sys
from typing import Dict, List, Tuple

# Load environment variables from .env file
try:
    from dotenv import load_dotenv
    
    # Determine environment (defaults to 'local')
    env = os.getenv('ENVIRONMENT', 'local').lower()
    
    # Load environment-specific .env file
    env_file = f'.env.{env}'
    if os.path.exists(env_file):
        load_dotenv(env_file)
        print(f"✓ Loaded environment from {env_file}")
    else:
        # Fallback to .env if environment-specific file doesn't exist
        load_dotenv('.env')
        print("✓ Loaded environment from .env")
except ImportError:
    print("⚠️  python-dotenv not installed. Install with: pip install python-dotenv")
except Exception as e:
    print(f"⚠️  Error loading .env file: {e}")

# Import broker classes from trading_bot
from trading_bot import SchwabBroker, RobinhoodBroker, BROKERS

# Color codes for terminal output
class Colors:
    GREEN = '\033[92m'
    RED = '\033[91m'
    YELLOW = '\033[93m'
    BLUE = '\033[94m'
    RESET = '\033[0m'
    BOLD = '\033[1m'

def print_success(message: str):
    """Print success message"""
    print(f"{Colors.GREEN}✓ {message}{Colors.RESET}")

def print_error(message: str):
    """Print error message"""
    print(f"{Colors.RED}✗ {message}{Colors.RESET}")

def print_warning(message: str):
    """Print warning message"""
    print(f"{Colors.YELLOW}⚠ {message}{Colors.RESET}")

def print_info(message: str):
    """Print info message"""
    print(f"{Colors.BLUE}ℹ {message}{Colors.RESET}")

def print_header(message: str):
    """Print header message"""
    print(f"\n{Colors.BOLD}{Colors.BLUE}{'='*60}{Colors.RESET}")
    print(f"{Colors.BOLD}{Colors.BLUE}{message}{Colors.RESET}")
    print(f"{Colors.BOLD}{Colors.BLUE}{'='*60}{Colors.RESET}\n")

def test_schwab_connection() -> Tuple[bool, Dict]:
    """Test Charles Schwab connection"""
    print_header("Testing Charles Schwab Connection")
    
    config = BROKERS['schwab']
    
    # Check if enabled
    if not config['enabled']:
        print_warning("Schwab is not enabled in configuration")
        print_info("Set SCHWAB_ENABLED=true in your .env file to test")
        return False, {}
    
    # Check credentials
    if not config['api_key'] or not config['app_secret']:
        print_error("Schwab API credentials not configured")
        print_info("Set SCHWAB_API_KEY and SCHWAB_APP_SECRET in your .env file")
        return False, {}
    
    print_info(f"API Key: {config['api_key'][:10]}... (hidden)")
    print_info(f"Callback URL: {config['callback_url']}")
    print_info(f"Token Path: {config['token_path']}")
    
    # Test connection
    try:
        broker = SchwabBroker(config)
        print_info("Attempting to connect to Charles Schwab API...")
        
        if broker.connect():
            print_success("Successfully connected to Charles Schwab API")
            
            # Test account info retrieval
            print_info("Testing account information retrieval...")
            account_info = broker.get_account_info()
            
            if account_info:
                print_success("Successfully retrieved account information")
                account_id = account_info.get('accountId', 'N/A')
                account_type = account_info.get('type', 'N/A')
                print_info(f"  Account ID: {account_id}")
                print_info(f"  Account Type: {account_type}")
                
                # Test positions retrieval
                print_info("Testing positions retrieval...")
                positions = broker.get_positions()
                print_success(f"Successfully retrieved {len(positions)} positions")
                
                # Test market data retrieval
                print_info("Testing market data retrieval (SPY)...")
                price, indicators = broker.get_market_data('SPY')
                if price > 0:
                    print_success(f"Successfully retrieved market data for SPY: ${price:.2f}")
                    print_info(f"  Bid: ${indicators.get('bid', 0):.2f}")
                    print_info(f"  Ask: ${indicators.get('ask', 0):.2f}")
                    print_info(f"  Volume: {indicators.get('volume', 0):,}")
                else:
                    print_warning("Market data retrieval returned no price")
                
                return True, {
                    'account_id': account_id,
                    'account_type': account_type,
                    'positions_count': len(positions),
                    'market_data_working': price > 0
                }
            else:
                print_warning("Connected but could not retrieve account information")
                return True, {'account_info': False}
        else:
            print_error("Failed to connect to Charles Schwab API")
            return False, {}
            
    except Exception as e:
        print_error(f"Error testing Schwab connection: {e}")
        import traceback
        print_info("Full error details:")
        traceback.print_exc()
        return False, {}

def test_robinhood_connection() -> Tuple[bool, Dict]:
    """Test Robinhood connection"""
    print_header("Testing Robinhood Connection")
    
    config = BROKERS['robinhood']
    
    # Check if enabled
    if not config['enabled']:
        print_warning("Robinhood is not enabled in configuration")
        print_info("Set ROBINHOOD_ENABLED=true in your .env file to test")
        return False, {}
    
    # Check credentials
    if not config['username'] or not config['password']:
        print_error("Robinhood credentials not configured")
        print_info("Set ROBINHOOD_USERNAME and ROBINHOOD_PASSWORD in your .env file")
        return False, {}
    
    print_info(f"Username: {config['username']}")
    print_info("Password: *** (hidden)")
    if config.get('mfa_code'):
        print_info("MFA Code: *** (hidden)")
    
    # Test connection
    try:
        broker = RobinhoodBroker(config)
        print_info("Attempting to connect to Robinhood API...")
        
        if broker.connect():
            print_success("Successfully connected to Robinhood API")
            
            # Test account info retrieval
            print_info("Testing account information retrieval...")
            account_info = broker.get_account_info()
            
            if account_info:
                print_success("Successfully retrieved account information")
                username = account_info.get('username', 'N/A')
                print_info(f"  Username: {username}")
                
                # Test positions retrieval
                print_info("Testing positions retrieval...")
                positions = broker.get_positions()
                print_success(f"Successfully retrieved {len(positions)} positions")
                
                # Test market data retrieval
                print_info("Testing market data retrieval (SPY)...")
                price, indicators = broker.get_market_data('SPY')
                if price > 0:
                    print_success(f"Successfully retrieved market data for SPY: ${price:.2f}")
                    print_info(f"  Bid: ${indicators.get('bid', 0):.2f}")
                    print_info(f"  Ask: ${indicators.get('ask', 0):.2f}")
                    print_info(f"  Volume: {indicators.get('volume', 0):,}")
                else:
                    print_warning("Market data retrieval returned no price")
                
                return True, {
                    'username': username,
                    'positions_count': len(positions),
                    'market_data_working': price > 0
                }
            else:
                print_warning("Connected but could not retrieve account information")
                return True, {'account_info': False}
        else:
            print_error("Failed to connect to Robinhood API")
            print_info("Check your username, password, and MFA code if enabled")
            return False, {}
            
    except Exception as e:
        print_error(f"Error testing Robinhood connection: {e}")
        import traceback
        print_info("Full error details:")
        traceback.print_exc()
        return False, {}

def main():
    """Main test function"""
    print_header("Trading Bot Connection Test")
    
    env = os.getenv('ENVIRONMENT', 'local').lower()
    print_info(f"Environment: {env}")
    print_info(f"Testing brokers based on configuration...\n")
    
    results = {
        'schwab': {'success': False, 'details': {}},
        'robinhood': {'success': False, 'details': {}}
    }
    
    # Test Schwab
    schwab_success, schwab_details = test_schwab_connection()
    results['schwab'] = {'success': schwab_success, 'details': schwab_details}
    
    # Test Robinhood
    robinhood_success, robinhood_details = test_robinhood_connection()
    results['robinhood'] = {'success': robinhood_success, 'details': robinhood_details}
    
    # Print summary
    print_header("Test Summary")
    
    if results['schwab']['success']:
        print_success("Charles Schwab: CONNECTED ✓")
        if results['schwab']['details']:
            details = results['schwab']['details']
            print_info(f"  Account ID: {details.get('account_id', 'N/A')}")
            print_info(f"  Positions: {details.get('positions_count', 0)}")
            print_info(f"  Market Data: {'Working' if details.get('market_data_working') else 'Not Working'}")
    else:
        if BROKERS['schwab']['enabled']:
            print_error("Charles Schwab: FAILED ✗")
        else:
            print_warning("Charles Schwab: DISABLED (not configured)")
    
    if results['robinhood']['success']:
        print_success("Robinhood: CONNECTED ✓")
        if results['robinhood']['details']:
            details = results['robinhood']['details']
            print_info(f"  Username: {details.get('username', 'N/A')}")
            print_info(f"  Positions: {details.get('positions_count', 0)}")
            print_info(f"  Market Data: {'Working' if details.get('market_data_working') else 'Not Working'}")
    else:
        if BROKERS['robinhood']['enabled']:
            print_error("Robinhood: FAILED ✗")
        else:
            print_warning("Robinhood: DISABLED (not configured)")
    
    # Exit code
    if results['schwab']['success'] or results['robinhood']['success']:
        print_success("\n✓ At least one broker connection successful!")
        sys.exit(0)
    else:
        print_error("\n✗ No broker connections successful. Please check your configuration.")
        sys.exit(1)

if __name__ == "__main__":
    main()

